src/ethereum_test_tools/common/types.py

```diff
@@ -489,13 +503,71 @@ def to_kwargs_dict(account: "Dict | Account | None") -> Dict:
         return cls(**kwargs)
 
 
+class Sender(Address):
+    """
+    Address of the sender of a transaction.
+    """
+
+    key: Hash | None
+    nonce: Number
+
+    def __new__(
+        cls,
+        address: "FixedSizeBytesConvertible | Address | Sender | None" = None,
+        *,
+        key: FixedSizeBytesConvertible | None = None,
+        nonce: NumberConvertible = 0,
+    ):
+        """
+        Init the sender.
+        """
+        if address is None:
+            if key is None:
+                raise ValueError("impossible to initialize sender without address")
+            private_key = PrivateKey(Hash(key))
+            public_key = private_key.public_key
+            address = Address(keccak256(public_key.format(compressed=False)[1:])[32 - 20 :])
+        elif isinstance(address, Sender):
+            return address
+        instance = super(Sender, cls).__new__(cls, address)
+        instance.key = Hash(key) if key is not None else None
+        instance.nonce = Number(nonce)
+        return instance
+
+    def copy(self) -> "Sender":
+        """
+        Returns a copy of the sender.
+        """
+        return Sender(Address(self), key=self.key, nonce=self.nonce)
+
+
+MAX_SENDERS = 50
+SENDERS_ITER = iter(
+    Sender(key=TestPrivateKey + i if i != 1 else TestPrivateKey2, nonce=0) for i in count(0)
+)
+SENDERS = [next(SENDERS_ITER) for _ in range(MAX_SENDERS)]
```

Comment:
Yes, this was simply done to not have to recalculate the address from the private key on each test, but I was thinking that we could optimize it like this: https://docs.python.org/3/library/functools.html#functools.cache

Reply:
This would be nice.

Reply:
But would be even nicer to avoid sender address collisions and generate a fresh set of addresses for every test, so I don't think the caching is actually that relevant. I added this in my first review below.

Reply:
Yes it would be nice to have a different sender on every test, but we would need to investigate how much of a bottle neck is the elliptic curve operations to recover the address for a new private key on every single test.

Reply:
Another thing about declaring a different sender for every test is that it would not be thread-safe, i.e. we would get different fixtures on every fill when using `fill -n auto`, and therefore the all-tests-in-a-single-block composed version of `fill` would not be able to be filled in parallel, hence we should only do this when we actually need a different sender for every test.